aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to cat {filename}: {output}')
+    return output
+
+
+def sed_replace(session, pattern, replacement, filename):
+    """
+    Replace a pattern in a file using sed, raising an error if nothing is found.
+
+    :param session: session of the remote host
+    :type session: :py:class:`RemoteSession`
+    :param str pattern: pattern to replace
+    :param str replacement: string to replace the pattern for
+    :param str filename: file with content to be replaced
+    :raises: :py:class:`AssertionError` if nothing was replaced
+
+    Using sed is convenient, but it does not error out when nothing is replaced
+    and this can lead to false-positives when used inside tests.
+    """
+    LOG.info("Replacing {pattern} with {replacement} in {filename}")
+    # for lines matching pattern, replace and print them (the p command won't work)
+    cmd = f"""sed -ri '/{pattern}/{{
+s//{replacement}/g
+w /dev/stdout
+}}' {filename}"""
+    output = session.cmd(cmd)
+    # make sure we have at least one line replaced
+    assert len(output.splitlines()) > 0, f"sed did not find any matches for pattern {pattern}"
+
+
+def grep(session, expr, path, check=False, flags="-a"):
+    """
+    Invoke ``grep`` remotely searching for an expression in a path.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str expr: search expression
+    :param str path: file to search
+    :param bool check: whether to quietly run grep for a boolean check
+    :param str flags: extra flags passed to ``grep`` on the command line
+    :returns: whether there is a match or not or what ``grep`` emits on stdout
+              if the check mode is disabled
+    :rtype: bool or str
+    :raises: ShellCmdError if the check mode is disabled and status is nonzero
+    """
+    grep_command = f"grep {flags} '{expr}' '{path}'"
+    status, output = session.cmd_status_output(grep_command)
+    if check:
+        return status == 0
+    if status != 0:
+        raise ShellCmdError(grep_command, status, output)
+    return output
+
+
+###############################################################################
+# utilities
+###############################################################################
+
+
+def hash_file(session, filename, method='md5'):
+    """
+    Calculate hash of given file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file that should be hashed
+    :param str method: method for hashing
+    :returns: hash of file, a hex number
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if hash command failed (e.g. file not
+             found) or resulting output does not have expected format
+    """
+    cmd = f"{method}sum"
+    expect_len = 32
```

Comment:
The expect_len should be only checked for `md5` as other cyphers are using different lengths (or we can build a map, but I don't think it'd be that necessary)

Reply:
Indeed, sounds a bit of an overkill and the length check is not that useful for the purpose of the function which is to generate a hash for a given file. Validating the hash characters should be enough in this case and still general enough.